internal/dag/cache.go

```diff
@@ -139,13 +157,39 @@ func (kc *KubernetesCache) Insert(obj interface{}) bool {
 	case *v1.Service:
 		kc.services[k8s.NamespacedNameOf(obj)] = obj
 		return kc.serviceTriggersRebuild(obj)
+	case *v1beta1.IngressClass:
+		// Match only ingress classes that Contour should be able to own.
+		if obj.Spec.Controller == "projectcontour.io/ingress-controller" {
+			kc.ingressclasses[obj.Name] = obj
+			return true
+		}
 	case *v1beta1.Ingress:
-		if kc.matchesIngressClass(obj) {
-			kc.ingresses[k8s.NamespacedNameOf(obj)] = obj
+		matches := kc.matchesIngressClass(obj.Annotations, obj.Spec.IngressClassName)
+		kc.ingresses[k8s.NamespacedNameOf(obj)] = obj
+
+		if matches {
 			return true
 		}
+
+		kind := k8s.KindOf(obj)
+		om := obj.GetObjectMeta()
+
+		ic := func() string {
+			if found, a := annotation.IngressClass(obj.Annotations); found {
+				return a
+			}
+			if obj.Spec.IngressClassName != nil {
+				return *obj.Spec.IngressClassName
+			}
+			return ""
+		}
```

Comment:
The problem is that we need to know if the annotation is defined at all, if it is then check if it matches. If it's not defined, then we can look at ingress class. That's why it's changed.

Reply:
I agree that having one function that returns an object's ingress class would be really helpful since the logic is spread around a bunch now. It may need to also return a bool indicating if the class came from Spec.IngressClassName or not.

Reply:
> The problem is that we need to know if the annotation is defined at all, if it is then check if it matches. If it's not defined, then we can look at ingress class. That's why it's changed.

I think that we can still have it in one place though. The logic for `IngressClass` would be:
```Go
func IngressClass(o runtime.Object) string {
	a := o.GetObjectMeta().GetAnnotations()
	if class, ok := a["projectcontour.io/ingress.class"]; ok {
		return class
	}
	if class, ok := a["contour.heptio.com/ingress.class"]; ok {
		return class
	}
	if class, ok := a["kubernetes.io/ingress.class"]; ok {
		return class
	}

        switch obj := obj.(type) {
        case *v1beta1.Ingress:
                return obj.Spec.IngressClassName
        }

        return ""
}
```

This treats the ingress class the same no matter how it is specified. Then we can treat the issue of whether the ingress class matches our configuration separately. We may not want to propagate the default Contour matching behaviour across the the IngressClass resource.